src/util.js

```diff
@@ -0,0 +1,66 @@
+import { getExecOutput } from '@actions/exec';
+import { debug, setFailed } from '@actions/core';
+import { HttpClient } from '@actions/http-client';
+import { createWriteStream } from 'fs';
+import { promisify } from 'util';
+import * as stream from 'stream';
+
+/**
+ * If caching may be used (more precisely, the git-set-file-times.pl script would be executed), check that the repo is not a shallow one, because if it is we will be missing time data
+ * @returns {Promise<void>}
+ */
+export async function checkRepositoryIsNotShallow() {
+    const workspace = getEnvOrThrow('GITHUB_WORKSPACE');
+    debug(`Checking that the repository checkout at ${workspace} is not shallow`);
+    const output = await getExecOutput('git', ['-C', workspace, 'rev-parse', '--is-shallow-repository'], {
+        silent: true
+    });
+    if (output.stdout === 'true\n') {
+        setFailed('The full commit history of the repository must be checked out. Please set the fetch-depth parameter of actions/checkout to 0.');
+    }
+}
+
+/**
+ * @returns {string}
+ */
+export function getArchitecture() {
+    return getEnvOrThrow('RUNNER_ARCH');
+}
+
+/**
+ * @return {string}
+ */
+export function getBranchName() {
+    const name = process.env['GITHUB_HEAD_REF'];
+    if (name) {
+        return name;
+    }
+    return process.env['GITHUB_REF_NAME'];
+}
```

Comment:
I like how this function looks! However, does it really return the branch name? As fair as I know, `GITHUB_REF_NAME` returns other things besides the branch name, and so may do `GITHUB_HEAD_REF`.

If not, maybe we can use these values as cache keys anyway (after all, we care they are unique per branch, not exactly that they are the branch name), but I think it'd be better to change the name of the function to reflect that.

Reply:
> `GITHUB_HEAD_REF`
> The head ref or source branch of the pull request in a workflow run. This property is only set when the event that triggers a workflow run is either pull_request or pull_request_target. For example, feature-branch-1.
>
> `GITHUB_REF_NAME`
> The branch or tag name that triggered the workflow run. For example, feature-branch-1.
>
> `GITHUB_REF_TYPE`
> The type of ref that triggered the workflow run. Valid values are branch or tag.

It seems that `GITHUB_REF_NAME` may be the tag name. We can rename `getBranchName` to `getRefName` or something, or get the branch name in another way when `GITHUB_REF_TYPE` is `tag`.

Reply:
Also, to increase the possibility of using the cache in PR, if it can not get the cache with `GITHUB_HEAD_REF`, it can try to use the cache with `GITHUB_BASE_REF`. What do you think?

> `GITHUB_BASE_REF`
> The name of the base ref or target branch of the pull request in a workflow run. This is only set when the event that triggers a workflow run is either pull_request or pull_request_target. For example, main.

Reply:
You're right about `GITHUB_REF_NAME`. I was mistaking it by the similar, but not the same `GITHUB_REF` variable :smile:

Taking into account what you've said and what we want to achieve, I think this plan would be great:

- We read the value of `GITHUB_REF_TYPE`:
    - If it is `branch`, continue with the next steps.
    - If it is `tag`, then disable caching, by making `restorePackSquashCache` do nothing and return `undefined`. I think this is the best solution at our disposal for several reasons:
        - Tags are glorified pointers to commits, so we don't know to what branch they belong, and tags by themselves do not represent a "history progression". It is okay for people to create the tags "green" and "red" without them having any relevant connection to each other, and sometimes this is done just to landmark points in the development.
        - Even if a tag was created previously and has a cache associated to it (in other words, we are modifying a tag), there is no guarantee that the commit history behind both versions of the tag did not diverge significantly.
        - It is sensible to assume that running PackSquash in workflows triggered only by tag events is so rare that caring about these concerns is worth the occasional slowdowns.
- `getBranchName` should work as follows:
    - We try to get the head branch name from `GITHUB_HEAD_REF`. If `GITHUB_HEAD_REF` contains a non-empty string (GitHub runners set it to an empty value outside of PRs), then we return that. Otherwise, keep trying the next steps.
    - We could try getting the base branch name from `GITHUB_BASE_REF`, but that risks reusing the same cache for development histories that have diverged too much, causing new files to be reused when in reality they changed. So I think it's best to only consider the head branch for PRs, and ignore `GITHUB_BASE_REF`. I've described this problem in more detail here (by the way, I think we should copy it to the new JavaScript code, as I find it pretty relevant still!): https://github.com/ComunidadAylas/PackSquash-action/blob/cd8aaf4b20f3634d35aba76a5e7b46acf0e92cc2/entrypoint.sh#L477-L498
    - We read `GITHUB_REF_NAME` and return that, as a branch name.

Do you agree with this plan? :smiling_face_with_three_hearts:

Reply:
Yes, of course!! Fix by 5be8b3d1e37bdd088feb9d3c663b2d924fd32690